lib/rex/post/meterpreter/packet.rb

```diff
@@ -693,17 +819,31 @@ def to_r
   # passing it on to the default functionality that can parse
   # the TLV values.
   #
-  def from_r(bytes)
-    xor_key = bytes[0,4].unpack('N')[0]
-    super(xor_bytes(xor_key, bytes[4, bytes.length]))
+  def from_r(aes_key=nil)
+    xor_key = self.raw[XOR_KEY_OFF, XOR_KEY_SIZE]
+    data = xor_key + xor_bytes(xor_key, self.raw[SESSION_GUID_OFF, self.raw.length - SESSION_GUID_OFF])
+    self.session_guid = data[SESSION_GUID_OFF, SESSION_GUID_SIZE]
+    encrypted_flag = data[ENCRYPTED_FLAG_OFF, ENCRYPTED_FLAG_SIZE] == "\x01"
+    packet_length = data[PACKET_LENGTH_OFF, PACKET_LENGTH_SIZE]
+    packet_type = data[PACKET_TYPE_OFF, PACKET_TYPE_SIZE]
+    self.type = packet_type.unpack('N')[0]
+    if encrypted_flag
```

Comment:
Is it an issue that we're trusting packets to essentially declare whether the session is under encryption? I think it allows a passive MitM to spoof TLV responses. Would "knowing" that a session is under encryption and that all TLV's must arrive encrypted provide a measure of client authentication? Will noodle if it's worth considering.